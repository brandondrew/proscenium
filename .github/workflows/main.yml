name: Ruby

on:
  push:
    branches:
      - master

  pull_request:

jobs:
  release:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v3

      - name: Set up Ruby
        uses: ruby/setup-ruby@v1
        with:
          ruby-version: 3.1

      - uses: denoland/setup-deno@v1.0.0
        with:
          deno-version: v1.x

      - name: Compile esbuild
        run: deno task compile:esbuild:macm

      - name: Install gem dependencies
        run: bundle install

      - name: Build native gem
        env:
          PROFILE: release
        run: bundle exec rake build

      - uses: actions/upload-artifact@v3
        with:
          name: artifacts
          path: bin/esbuild-*
