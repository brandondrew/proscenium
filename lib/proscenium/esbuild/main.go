package main

// go build -buildmode=c-shared -o main.so main.go

import "C"
import "os"
import "strings"
import "encoding/json"

// import "fmt"
import "github.com/evanw/esbuild/pkg/api"

//export transform
func transform(input *C.char) *C.char {
	inputString := C.GoString(input)
	result := api.Transform(inputString, api.TransformOptions{
		Loader: api.LoaderJS,
	})

	s := string(result.Code)

	return C.CString(s)
}

var envPlugin = api.Plugin{
	Name: "env",
	Setup: func(build api.PluginBuild) {
		// Intercept import paths called "env" so esbuild doesn't attempt
		// to map them to a file system location. Tag them with the "env-ns"
		// namespace to reserve them for this plugin.
		build.OnResolve(api.OnResolveOptions{Filter: `^env$`},
			func(args api.OnResolveArgs) (api.OnResolveResult, error) {
				return api.OnResolveResult{
					Path:      args.Path,
					Namespace: "env-ns",
				}, nil
			})

		// Load paths tagged with the "env-ns" namespace and behave as if
		// they point to a JSON file containing the environment variables.
		build.OnLoad(api.OnLoadOptions{Filter: `.*`, Namespace: "env-ns"},
			func(args api.OnLoadArgs) (api.OnLoadResult, error) {
				mappings := make(map[string]string)
				for _, item := range os.Environ() {
					if equals := strings.IndexByte(item, '='); equals != -1 {
						mappings[item[:equals]] = item[equals+1:]
					}
				}
				bytes, err := json.Marshal(mappings)
				if err != nil {
					return api.OnLoadResult{}, err
				}
				contents := string(bytes)
				return api.OnLoadResult{
					Contents: &contents,
					Loader:   api.LoaderJSON,
				}, nil
			})
	},
}

//export build
func build(entrypoint *C.char) *C.char {
	entrypointString := C.GoString(entrypoint)

	result := api.Build(api.BuildOptions{
		EntryPoints: []string{entrypointString},
		Bundle:      true,
		Plugins:     []api.Plugin{envPlugin},
		Write:       false,
	})

	outputString := string(result.OutputFiles[0].Contents)
	return C.CString(outputString)
}

func main() {
	// fmt.Println(transform())
	// fmt.Printf("%s", build("input.ts"))
}
