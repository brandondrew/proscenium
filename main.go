package main

/*
struct Result {
	int success;
	char* response;
};
*/
import "C"

import (
	"encoding/json"
	"joelmoss/proscenium/internal/builder"
	"joelmoss/proscenium/internal/resolver"
	"joelmoss/proscenium/internal/types"
	"path"
	"strings"
)

// Build the given `path` in the `root`.
//
//   - path - The path to build relative to `root`.
//   - root - The working directory.
//   - env - The environment (1 = development, 2 = test, 3 = production)
//   - importMap - Path to the import map relative to `root`.
//   - debug
//
//export build
func build(filepath *C.char, root *C.char, env C.uint, importMap *C.char, debug bool) C.struct_Result {
	types.Env = types.Environment(env)

	pathStr := C.GoString(filepath)

	result := builder.Build(builder.BuildOptions{
		Path:          pathStr,
		Root:          C.GoString(root),
		ImportMapPath: C.GoString(importMap),
		Debug:         debug,
	})

	if len(result.Errors) != 0 {
		j, err := json.Marshal(result.Errors[0])
		if err != nil {
			return C.struct_Result{C.int(0), C.CString(string(err.Error()))}
		}

		return C.struct_Result{C.int(0), C.CString(string(j))}
	}

	contents := string(result.OutputFiles[0].Contents)

	isSourceMap := strings.HasSuffix(pathStr, ".map")
	if isSourceMap {
		return C.struct_Result{C.int(1), C.CString(contents)}
	}

	sourcemapUrl := path.Base(pathStr)
	if strings.HasSuffix(result.OutputFiles[0].Path, ".css") {
		contents += "/*# sourceMappingURL=" + sourcemapUrl + ".map */"
	} else {
		contents += "//# sourceMappingURL=" + sourcemapUrl + ".map"
	}

	return C.struct_Result{C.int(1), C.CString(contents)}
}

// Resolve the given `path` relative to the `root`.
//
//   - path - The path to build relative to `root`.
//   - root - The working directory.
//   - env - The environment (1 = development, 2 = test, 3 = production)
//   - importMap - Path to the import map relative to `root`.
//
//export resolve
func resolve(path *C.char, root *C.char, env C.uint, importMap *C.char) C.struct_Result {
	types.Env = types.Environment(env)

	resolvedPath, err := resolver.Resolve(resolver.Options{
		Path:          C.GoString(path),
		Root:          C.GoString(root),
		ImportMapPath: C.GoString(importMap),
	})
	if err != nil {
		return C.struct_Result{C.int(0), C.CString(string(err.Error()))}
	}

	return C.struct_Result{C.int(1), C.CString(resolvedPath)}
}

func main() {}
