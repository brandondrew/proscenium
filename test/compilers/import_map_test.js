import { assertStringIncludes } from 'std/testing/asserts.ts'
import { assertSnapshot } from 'std/testing/snapshot.ts'
import { join } from 'std/path/mod.ts'
import { beforeEach, describe, it } from 'std/testing/bdd.ts'

import main from '../../lib/proscenium/compilers/esbuild.js'

const root = join(Deno.cwd(), 'test', 'internal')
const lightningcssBin = join(Deno.cwd(), 'bin', 'lightningcss')

describe('import map', () => {
  beforeEach(() => {
    Deno.env.set('RAILS_ENV', 'test')
    Deno.env.set('PROSCENIUM_TEST', 'test')
  })

  it('from json', async t => {
    const result = await main('lib/import_map.js', {
      root,
      lightningcssBin,
      importMap: 'config/import_maps/as.json'
    })

    await assertSnapshot(t, new TextDecoder().decode(result))
  })

  it('from js', async () => {
    const result = await main('lib/import_map_as_js.js', {
      root,
      lightningcssBin,
      importMap: 'config/import_maps/as.js'
    })

    assertStringIncludes(new TextDecoder().decode(result), 'import pkg from "/lib/foo2.js";')
  })

  it('maps imports via trailing slash', async () => {
    const result = await main('lib/component.jsx', {
      root,
      lightningcssBin,
      importMap: 'config/import_maps/trailing_slash_import.json'
    })

    assertStringIncludes(
      new TextDecoder().decode(result),
      'import { jsx } from "/url:https%3A%2F%2Fesm.sh%2Freact%4018.2.0%2Fjsx-runtime"'
    )
  })

  it('resolves imports from a node_module', async t => {
    const result = await main('node_modules/is-ip/index.js', {
      root,
      lightningcssBin,
      importMap: 'config/import_maps/npm.json'
    })

    await assertSnapshot(t, new TextDecoder().decode(result))
  })

  it('supports scopes', async t => {
    const result = await main('lib/import_map/scopes.js', {
      root,
      lightningcssBin,
      importMap: 'config/import_maps/scopes.json'
    })

    await assertSnapshot(t, new TextDecoder().decode(result))
  })

  it('supports aliasing', async t => {
    const result = await main('lib/import_map/aliases.js', {
      root,
      lightningcssBin,
      importMap: 'config/import_maps/aliases.json'
    })

    await assertSnapshot(t, new TextDecoder().decode(result))
  })
})
