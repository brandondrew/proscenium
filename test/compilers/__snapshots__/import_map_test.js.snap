export const snapshot = {};

snapshot[`import map > from json 1`] = `
'// lib/import_map.js
import axios from "/url:https%3A%2F%2Fcdnjs.cloudflare.com%2Fajax%2Flibs%2Faxios%2F0.24.0%2Faxios.min.js";
import { createElement } from "/url:https%3A%2F%2Fesm.sh%2Freact%4018.2.0";
import foo from "/lib/foo.js";
axios();
foo();
createElement();
import("/url:https%3A%2F%2Fesm.sh%2Fsolid%400.2.1");
//# sourceMappingURL=import_map.js.map
'
`;

snapshot[`import map > resolves imports from a node_module 1`] = `
'// node_modules/.pnpm/is-ip@5.0.0/node_modules/is-ip/index.js
import ipRegex from "/url:https%3A%2F%2Fesm.sh%2Fip-regex";
import { isMatch } from "/node_modules/.pnpm/super-regex@0.2.0/node_modules/super-regex/index.js";
var maxIPv4Length = 15;
var maxIPv6Length = 45;
var options = {
  timeout: 400
};
function isIP(string) {
  return isMatch(ipRegex({ exact: true }), string.slice(0, maxIPv6Length), options);
}
function isIPv6(string) {
  return isMatch(ipRegex.v6({ exact: true }), string.slice(0, maxIPv6Length), options);
}
function isIPv4(string) {
  return isMatch(ipRegex.v4({ exact: true }), string.slice(0, maxIPv4Length), options);
}
function ipVersion(string) {
  return isIP(string) ? isIPv6(string) ? 6 : 4 : void 0;
}
export {
  ipVersion,
  isIP,
  isIPv4,
  isIPv6
};
//# sourceMappingURL=index.js.map
'
`;

snapshot[`import map > supports scopes 1`] = `
'// lib/import_map/scopes.js
import "/lib/foo2.js";
//# sourceMappingURL=scopes.js.map
'
`;

snapshot[`import map > supports aliasing 1`] = `
'// lib/import_map/aliases.js
import isip from "/node_modules/.pnpm/is-ip@5.0.0/node_modules/is-ip/index.js";
isip();
//# sourceMappingURL=aliases.js.map
'
`;
